odoo/addons/base/models/res_users.py

```diff
@@ -1098,16 +1098,18 @@ def _apply_group(self, implied_group):
         """ Add the given group to the groups implied by the current group
         :param implied_group: the implied group to add
         """
-        if implied_group not in self.implied_ids:
-            self.write({'implied_ids': [Command.link(implied_group.id)]})
+        for group in self:
+            if implied_group not in group.implied_ids:
+                group.write({'implied_ids': [Command.link(implied_group.id)]})
 
     def _remove_group(self, implied_group):
         """ Remove the given group from the implied groups of the current group
         :param implied_group: the implied group to remove
         """
-        if implied_group in self.implied_ids:
-            self.write({'implied_ids': [Command.unlink(implied_group.id)]})
-            implied_group.write({'users': [Command.unlink(user.id) for user in self.users]})
+        for group in self:
+            if implied_group in group.implied_ids:
+                group.write({'implied_ids': [Command.unlink(implied_group.id)]})
+                implied_group.write({'users': [Command.unlink(user.id) for user in group.users]})
```

Comment:
Same as above.

Reply:
Note that although
```py
if any(implied_group not in group.implied_ids for group in self):
    self.write({'implied_ids': [Command.link(implied_group.id)]})
```
seems correct for `_appply_group`, the below alternative for `_remove_group`
```py
if any(implied_group in group.implied_ids for group in self):
    self.write({'implied_ids': [Command.unlink(implied_group.id)]})
    implied_group.write({'users': [Command.unlink(user.id) for user in self.users]})
```
seems odd, in the sense that we may remove more users than expected. We could write the last line as:
```py
    implied_group.write({'users': [Command.unlink(user.id) for user in self.filtered(lambda g: implied_group in g.implied_ids).users]})
```
but then we'll be using `filtered` anyway.

Form the (little) usage of this functions in current code, I would expect `self` to be a rather small recordset (few with more than two groups[^1]).

[^1]: One example with three groups https://github.com/odoo/odoo/blob/1c64e96ec683e3f28028031cdcfa6ae879ff6a6a/addons/website_sale/models/res_config_settings.py#L21